src/vsr/replica.zig

```diff
@@ -1400,46 +1450,105 @@ pub fn ReplicaType(
+            assert(self.view_headers.array.get(0).view <= self.view);
+            assert(self.view_headers.array.get(0).op == message.header.op);
+
+            {
+                const primary_repair_min =
+                    self.view_headers.array.get(self.view_headers.array.len - 1).op;
+                var op = self.commit_min + 1;
+
+                // Use "op ≤ primary_repair_min" instead of "=" to force the logs to intersect.
+                // This will cause a state transfer in the case where:
+                // - the repairing replica has a pristine log up to its op_checkpoint_trigger=X
+                // - the SV headers begin at op X+1
+                // But in exchange it avoids a special-case to update the commit_max and verify
+                // X's header.
+                const repairable = while (op <= primary_repair_min) : (op += 1) {
+                    const slot = self.journal.slot_with_op(op) orelse break false;
+                    if (self.journal.dirty.bit(slot)) break false;
+                } else true;
+
+                if (!repairable) {
+                    // This replica is too far behind, i.e. the new `self.op` is too far ahead of
+                    // the last checkpoint. If we wrap now, we overwrite un-checkpointed transfers
+                    // in the WAL, precluding recovery.
+                    // TODO State transfer.
+                    @panic("unimplemented (state transfer)");
+                }
             }
 
-            assert(self.op == message.header.op);
+            for (self.view_headers.array.constSlice()) |*header| {
+                if (header.op <= self.op_checkpoint_trigger()) {
+                    if (self.log_view < self.view or
+                        (self.log_view == self.view and header.op >= self.op))
+                    {
+                        self.set_op_and_commit_max(header.op, message.header.commit, "on_start_view");
+                        assert(self.op == header.op);
+                        assert(self.commit_max >= message.header.commit);
+                        break;
+                    }
+                }
+            } else {
+                log.err("{}: on_start_view: missing checkpoint trigger " ++
+                    "(view={} op_checkpoint_trigger={} op_head={})", .{
+                    self.replica,
+                    self.view,
+                    self.op_checkpoint_trigger(),
+                    self.op,
+                });
 
-            assert(self.status == .view_change);
-            self.transition_to_normal_from_view_change_status(message.header.view);
-            self.send_prepare_oks_after_view_change();
+                for (self.view_headers.array.constSlice()) |*header| {
+                    log.err("{}: on_start_view: header={}", .{ self.replica, header.* });
+                }
+                unreachable;
+            }
```

Comment:
In this branch, according to the SV, the primary could provide the checkpoint-trigger header that we need in order to repair our journal — but for some reason did not include it in the SV.
This _should_ be unreachable — primaries should always include the trigger header.

But from a Postel's law perspective, maybe we should ignore this (error message then return, to wait for another SV), or even fall back to state transfer.